MM-17912: Allow searching for files through plugin API

ashishbhate opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->
Allow searching for files through plugin API

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/12699

Otherwise, link the JIRA ticket.
-->
Fixes https://github.com/mattermost/mattermost-server/issues/12699
https://mattermost.atlassian.net/browse/MM-17912

ashishbhate commented on store/sqlstore/file_info_store.go:
```diff
@@ -95,6 +96,72 @@ func (fs SqlFileInfoStore) Get(id string) (*model.FileInfo, *model.AppError) {
+func (fs SqlFileInfoStore) GetWithOptions(opt *model.GetFilesOptions) ([]*model.FileInfo, *model.AppError) {
+	if opt == nil {
+		opt = &model.GetFilesOptions{}
+	}
```
`nil` options is valid because the zero value of `model.GetFilesOptions` is valid. It gets all files except the deleted ones. I think is a reasonable default.

I code does some basic validity checks of the options passed in, but it might better to have a method to check if the options are valid.

ashishbhate commented on store/sqlstore/file_info_store.go:
```diff
@@ -95,6 +96,72 @@ func (fs SqlFileInfoStore) Get(id string) (*model.FileInfo, *model.AppError) {
+		if opt.Page > 0 {
+			query = query.Offset(uint64(opt.PerPage * (opt.Page - 1)))
+		}
+	}
```
Pagination with out ordering will give inconsistent results. 

Maybe return an error if pagination is requested without ordering?

lieut-data replied:
See above for discussion on this, but by forcing `PerPage` to be required, we can check for `PerPage == 0` and just short-circuit the entire query. Should also verify that `Page >= 0` otherwise throw an error.

ashishbhate commented:
The master branch has progressed a lot since this PR was opened so one of the checks is failing. I'll merge or rebase when doing review fixes.

lieut-data commented on plugin/api.go:
```diff
@@ -659,6 +659,12 @@ type API interface {
+	// GetFileInfos gets File Infos with options
+	//
+	// @tag File
+	// Minimum server version: 5.20
```
This will now be:
```
	// Minimum server version: 5.22
```

lieut-data commented on model/file.go:
```diff
@@ -22,6 +22,36 @@ type FileUploadResponse struct {
+)
+
+// GetFilesOptions contains options for getting files
+type GetFilesOptions struct {
```
I know this name comes right out of the ticket, but I wonder if we should call this `GetFileInfoOptions` to match the terminology elsewhere.

ashishbhate replied:
I wasn't sure of the name either. I decided to use `GetFilesOptions` instead of `GetFileInfoOptions` because I thought this could potentially be reused for other file query methods.

ashishbhate replied:
I'll defer to whatever you think is appropriate.

lieut-data replied:
I'd be inclined to go with `FileInfo` as the underlying noun here, as it helps emphasize that this is the metadata and not the files themselves.

lieut-data commented on model/file.go:
```diff
@@ -22,6 +22,36 @@ type FileUploadResponse struct {
+// GetFilesOptions contains options for getting files
+type GetFilesOptions struct {
+	// UserIds optionally limits the files to those created by the given users.
+	UserIds []string
```
Could we define `json` struct field tag in case we need to serialize this for any reason? (i.e. `json:"user_ids"`)

ashishbhate replied:
I wasn't sure if this would require serialization. I'll add json tags.

lieut-data commented on model/file.go:
```diff
@@ -22,6 +22,36 @@ type FileUploadResponse struct {
+	Since int64
+	// IncludeDeleted includes deleted files if set.
+	IncludeDeleted bool
+	// Page optionally limits to the requested page of results.
```
Let's promote `Page` and `PerPage` to be explicit parameters passed to the method, i.e. `GetFileInfos(page, perPage, options)`. We hadn't arrived at this conclusion when the first ticket was filed, but we don't want to support `GetAll`-type of methods, but force paging to limit the volume of reads that can occur. If we leave this in the `options` struct and instead set some defaults for paging, a developer might think that page was the full set of results without ever writing the necessary paging code.

lieut-data commented on model/file.go:
```diff
@@ -22,6 +22,36 @@ type FileUploadResponse struct {
+	Page int
+	// PerPage optionally limits the number of results to fetch per page.
+	PerPage int
+	// SortBy optionally sorts the files by this field.
```
Can we specify a default sort order for when this string is left empty?

ashishbhate replied:
Sounds good, how about sorted by created at?

lieut-data replied:
Sounds good!

lieut-data commented on model/file.go:
```diff
@@ -22,6 +22,36 @@ type FileUploadResponse struct {
+	PerPage int
+	// SortBy optionally sorts the files by this field.
+	SortBy string
+	// SortDirection sorts the files in a particular order, default is ASCENDING if SortBy is provided.
```
The constant `FILE_SORT_ORDER_ASCENDING` is just `ASC`, not `ASCENDING` which might be confusing. To my surprise, we don't have any convention in the code to unify this. Should we define a reusable Go type to enforce this?
```go
type SortDirection int
const (
    Ascending SortDirection = 0
    Descending SortDirection = 1
)
```

I pick `int` only because its default value is "ready to go" vs. having to handle the empty string. 0/5.

ashishbhate replied:
I used strings here because the switch case in the query code becomes less nested because I don't have to set the sql string based on a number comparison.

Defining a type is good because it can enforce some compile time checks (and helps with code completion).

How about `byte` instead?

lieut-data replied:
I'd be inclined to stick with an `int` vs. a `byte` -- the latter is more compact, but typically also signals a different semantic meaning.

I do appreciate your point re: building the SQL query, but suspect that we can just wrap this with a helper method as needed. One potential thought to my original proposal is that it would be cool if `Ascending == -Descending`, for when you want to have a `sortDirection` and easily invert it. Maybe instead of the complexity of a type, we just have a `SortAscending bool` field, solving all the pain points?

ashishbhate replied:
>...we just have a SortAscending bool field, solving all the pain points?

This leads to the "hidden" implication that if `SortAscending` is not set, then the results are ordered in descending order. I will add a note in the docs, so I'm fine with this, but it be a gotcha to people who dont read the docs since its no explicit enough.

lieut-data replied:
Oops, you're right: what about `SortDescending` instead?

ashishbhate replied:
Sounds good to me.

I'll ping you once the PR is ready for a second round of reviews.

lieut-data commented on store/sqlstore/file_info_store.go:
```diff
@@ -95,6 +96,72 @@ func (fs SqlFileInfoStore) Get(id string) (*model.FileInfo, *model.AppError) {
+	query := fs.getQueryBuilder().
+		Select("FileInfo.*").
+		From("FileInfo").
+		LeftJoin("Posts ON FileInfo.PostId = Posts.Id") // Left join to get files that aren't attached to a post
```
We only use this JOIN in two cases below -- can we make it conditional as such?

lieut-data commented on store/sqlstore/file_info_store.go:
```diff
@@ -95,6 +96,72 @@ func (fs SqlFileInfoStore) Get(id string) (*model.FileInfo, *model.AppError) {
+	}
+
+	if opt.Since > 0 {
+		query = query.Where(sq.Gt{"FileInfo.CreateAt": opt.Since})
```
Should we use `Gte` here for a more natural definition of `Since`?

lieut-data commented on store/sqlstore/file_info_store.go:
```diff
@@ -95,6 +96,72 @@ func (fs SqlFileInfoStore) Get(id string) (*model.FileInfo, *model.AppError) {
+	if opt.SortDirection != "" {
+		sortDirection = opt.SortDirection
+	}
+	if opt.SortBy != "" {
```
Propose we pick `CreateAt` as the default when empty.

lieut-data commented on store/sqlstore/file_info_store.go:
```diff
@@ -95,6 +96,72 @@ func (fs SqlFileInfoStore) Get(id string) (*model.FileInfo, *model.AppError) {
+	if opt.SortBy != "" {
+		switch opt.SortBy {
+		case model.FILE_SORT_BY_CREATED, model.FILE_SORT_BY_SIZE:
+			query = query.OrderBy(fmt.Sprintf("FileInfo.%s %s", opt.SortBy, sortDirection))
```
Let's make all sorts stable by adding a secondary sort on `Id` for when two file infos have the same property as the primary sort.

lieut-data replied:
Also, can we write out the `OrderBy` explicitly for these two fields instead of using `fmt.Sprintf`? I think the code would be clearer as a result.

lieut-data commented on store/sqlstore/file_info_store.go:
```diff
@@ -95,6 +96,72 @@ func (fs SqlFileInfoStore) Get(id string) (*model.FileInfo, *model.AppError) {
+		switch opt.SortBy {
+		case model.FILE_SORT_BY_CREATED, model.FILE_SORT_BY_SIZE:
+			query = query.OrderBy(fmt.Sprintf("FileInfo.%s %s", opt.SortBy, sortDirection))
+		case model.FILE_SORT_BY_USERNAME:
```
I'm struggling to think of a case where this sort would be useful, and the join cost is non-trivial here. I'd suggest skipping this one.

ashishbhate replied:
I wasn't sure of this either.  

Since the caller is able to filter by user id and channel id, I thought the caller might want to sort by the names as well (because sorting by id didn't seem useful).

I'll defer to your call.

lieut-data replied:
I propose we remove this for now -- easy to revisit if we can come up with a compelling scenario, but simpler code is always easier to maintain and test.

lieut-data commented on store/sqlstore/file_info_store.go:
```diff
@@ -95,6 +96,72 @@ func (fs SqlFileInfoStore) Get(id string) (*model.FileInfo, *model.AppError) {
+		case model.FILE_SORT_BY_USERNAME:
+			query = query.Join("Users on FileInfo.CreatorId = Users.Id").
+				OrderBy(fmt.Sprintf("Users.%s %s", opt.SortBy, sortDirection))
+		case model.FILE_SORT_BY_CHANNEL_NAME:
```
I'm struggling to think of a case where this sort would be useful, and the join cost is non-trivial here. I'd suggest skipping this one.

ashishbhate commented:
@lieut-data thanks for the review. I've added a :+1:  to the comments that I have fixed locally. I'll push these changes (along with others) after we've resolved the open discussions.

lieut-data commented:
Thanks, @ashishbhate! I've commented above -- feel free to ping me again when you want me to take a second look at all the changes :)

ashishbhate commented:
@lieut-data The PR is ready for another round of reviews

lieut-data commented:
Terrific! I've re-queued myself for review and will get you feedback soon :)

lieut-data commented on model/file_info.go:
```diff
@@ -15,6 +15,27 @@ import (
+	UserIds []string `json:"user_ids"`
+	// ChannelIds optionally limits the FileInfos to those created in the given channels.
+	ChannelIds []string `json:"channel_ids"`
+	// Since optionally limits FileInfos to those created after the given time, specified as Unix time in milliseconds.
```
minor tweak to reflect code change:
```suggestion
	// Since optionally limits FileInfos to those created at or after the given time, specified as Unix time in milliseconds.
```

ashishbhate replied:
Good catch, I completely missed this change when I updated the code.

lieut-data commented on model/file_info.go:
```diff
@@ -15,6 +15,27 @@ import (
+	IncludeDeleted bool `json:"include_deleted"`
+	// SortBy sorts the FileInfos by this field. The default is to sort by date created.
+	SortBy string `json:"sort_by"`
+	// SortDescending if set sorts the FileInfos in descending order. The default is to sort in ascending order.
```
Let's simplify:
```suggestion
	// SortDescending changes the sort direction to descending order when true.
```

lieut-data commented on store/sqlstore/file_info_store.go:
```diff
@@ -95,6 +95,72 @@ func (fs SqlFileInfoStore) Get(id string) (*model.FileInfo, *model.AppError) {
 }
 
+func (fs SqlFileInfoStore) GetWithOptions(page, perPage uint, opt *model.GetFileInfosOptions) ([]*model.FileInfo, *model.AppError) {
+	if perPage == 0 || page == 0 {
```
I think `page == 0` should be a valid input, since we use 0 based semantics everywhere else.

ashishbhate replied:
I noticed that we use `int` everywhere else for `page` and `perPage`. I've used `uint` here as a "free" compile time check. Postgres and MySql don't allow negative offset or limit. I think a long time ago Postgres would silently take negative offsets as zero.

Should I change to `int` to be consistent with other parts of the code?

lieut-data replied:
Yes, that's an excellent call -- let's switch to `int`.

lieut-data commented on store/sqlstore/file_info_store.go:
```diff
@@ -95,6 +95,72 @@ func (fs SqlFileInfoStore) Get(id string) (*model.FileInfo, *model.AppError) {
+		return nil, nil
+	}
+
+	// Nil option means sort by create at ascending
```
nit: this comment seems misplaced. We might have been given a `&model.GetFileInfosOptions{}` for which the sort is still undefined but this code block doesn't deal with that. Suggest we remove and leave to the later code.
```suggestion
```

ashishbhate replied:
The intent was to say that a zero-value options struct, passed in either as a nil pointer or a pointer to an zero-value struct, is not an error case and sensible defaults are applied. I guess this is evident from the code so dropping the comment sounds good to me.

lieut-data commented on store/storetest/file_info_store.go:
```diff
@@ -256,6 +258,133 @@ func testFileInfoGetForUser(t *testing.T, ss store.Store) {
+			ExpectedFileIds: []string{file1_1.Id, file1_2.Id, file1_3.Id, file2_1.Id, file2_2.Id},
+		},
+		{
+			Name:    "Get files including deleted filtered by channel",
```
Can we also test filtering by user?

ashishbhate replied:
There was already a test for filtering by user a few lines below. I've added another test to test filtering by user and channel.

lieut-data replied:
Whoops! I thought I looked carefully, but perhaps not -- thanks for the bonus test anyway :)

ashishbhate commented:
@lieut-data I've updated based on your review feedback. Just one question about `int` vs `uint` above.

ashishbhate commented:
I've removed the unnecessary whitespace and changed the code to use `int` instead of `uint`

Thanks for your patience with the reviews!

lieut-data commented:
@prapti, testing this will require writing custom plugin. We likely want to extend the demo plugin to illustrate, perhaps with a simple `/search_files` slash command or some such.

ashishbhate commented:
@lieut-data If we want go down the route of a slash command in the demo plugin as an illustration, I'd be happy to implement it.

I'm not sure exactly what you had in mind, so a rough spec to get started would be great.

lieut-data commented:
@ashishbhate, thanks for offering to help further here! We primarily use the demo plugin to enable verification of features added to the API, with a secondary goal of illustrating how to do it from a code perspective. But the actual functionality is usually just "made up" on the spot to best illustrate and not necessarily be a cohesive experience like a traditional plugin.

In this case, I was imagining a new slash command -- `/list_files` registered by the demo plugin and using this API to post a new message in the current channel with the last 10 files attached to posts. A more elaborate approach might be to showcase those files in a custom RHS that included infinite scrolling, etc.

Let me know your appetite here, and I'd be happy to craft an appropriate help-wanted ticket to send your way, or you could just include it as part of this effort.

ashishbhate commented:
@lieut-data Adding a slash command that posts a new message listing files should be easy. I'm not yet sure about the custom RHS, I haven't implemented one before so I'm not sure of what it involves exactly. 

I'll open a PR in the demo plugin repo over the weekend to at least post a file list to a channel. We can then build on top of that based on feedback.

Lets consider the demo plugin part as a continuation of this effort. Fewer things to keep track of ;)

hanzei commented on app/file.go:
```diff
@@ -1092,6 +1092,10 @@ func (a *App) GetFileInfo(fileId string) (*model.FileInfo, *model.AppError) {
 	return a.Srv.Store.FileInfo().Get(fileId)
 }
 
+func (a *App) GetFileInfos(page, perPage int, opt *model.GetFileInfosOptions) ([]*model.FileInfo, *model.AppError) {
```
One comment on the API design: What does the API return in `nil` as options is provided?

ashishbhate replied:
FileInfos that are not deleted, sorted in ascending order of date created.

hanzei replied:
Thanks for the answer @ashishbhate. That should be sufficed :+1:

hanzei replied:
Would it make the API more clear if a struct is expected instead of a pointer to a struct?

ashishbhate replied:
I don't think switching to a struct instead of pointer makes the API any clearer. They're both nearly equivalent and there is existing convention in the code base to pass options as pointers.

hanzei replied:
Thanks for sharing your thoughts @ashishbhate :+1:

ashishbhate commented:
I've opened a PR that implements simple file listing in the demo-plugin

https://github.com/mattermost/mattermost-plugin-demo/pull/88

I haven't fully figured out the frontend changes required to implement an RHS panel.

lieut-data commented:
/update-branch

aaronrothschild commented:
Checked that this is Documented for 5.22 here: https://developers.mattermost.com/extend/plugins/server/reference/#API.GetFileInfos
